src/shims/foreign_items.rs

```diff
@@ -185,25 +220,31 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
 
         // Third: functions that return.
-        if this.emulate_foreign_item_by_name(link_name, abi, args, dest, ret)? {
-            trace!("{:?}", this.dump_place(**dest));
-            this.go_to_block(ret);
+        if let Some(returns) = this.emulate_foreign_item_by_name(link_name, abi, args, dest, ret)? {
+            if returns {
+                trace!("{:?}", this.dump_place(**dest));
+                this.go_to_block(ret);
+            }
+        } else if let Some(body) = this.lookup_non_external_foreign_item(abi, link_name_sym)? {
```

Comment:
Note that this code will never be reached for functions without a return place. But is it even possible to write `-> !` functions in `extern` blocks?

Reply:
It's possible but I didn't notice there's a `throw_unsup_format!("can't call (diverging) foreign function: {}", link_name)` above. 😰
```
error: unsupported operation: can't call (diverging) foreign function: no_return
  --> $DIR/non_extern_foreign_call_no_return.rs:10:14
   |
10 |     unsafe { no_return() }
   |              ^^^^^^^^^^^ can't call (diverging) foreign function: no_return
   |
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
```

Reply:
Yeah that is needed to avoid ICEs from later `unwrap`s.